node/src/types/block.rs

```diff
@@ -1265,13 +1266,38 @@ impl Display for BlockSignatures {
 
 /// A proto-block after execution, with the resulting post-state-hash.  This is the core component
 /// of the Casper linear blockchain.
-#[derive(DataSize, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[derive(DataSize, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Hash, Serialize)]
 pub struct Block {
     hash: BlockHash,
     header: BlockHeader,
     body: BlockBody,
 }
 
+/// A temporary copy of a block that has not been validated yet.
+#[derive(Debug, Deserialize)]
+pub(crate) struct UnverifiedBlock {
+    pub(crate) hash: BlockHash,
+    pub(crate) header: BlockHeader,
+    pub(crate) body: BlockBody,
+}
+
+impl TryFrom<UnverifiedBlock> for Block {
+    type Error = BlockValidationError;
+
+    fn try_from(unverified_block: UnverifiedBlock) -> Result<Block, BlockValidationError> {
```

Comment:
My preference is to not put domain logic in `TryFrom`s and instead have a smart constructor like `Block::from_unverified(unverified_block)` that would perform the verification. `fn verify` doesn't have to be a method on `Block` object - it can be a standalone function.

Reply:
I turned it into `UnverifiedBlock::verify` instead: https://github.com/casper-network/casper-node/pull/1928/commits/e7bc5da9bd367ed5967ec5e1cc649649875c3495
Not sure if that's better. I think I'd prefer `TryFrom`, but I'm fine with either.

Reply:
I would still make it a smart constructor - conceptually, an `UnverifiedBlock` shouldn't know how to "become" verified. It's the responsibility of the `Block` to turn an unverified one into verified. This also means that anyone having an instance of `UnverifiedObject` could call potentially expensive `verify`. If `Block::from_unverified` were to live in a smaller scope then we could make sure that it's only ever called once and we never have an instance of `UnverifiedBlock` anywhere.

Reply:
Done in https://github.com/casper-network/casper-node/pull/1928/commits/d14c0eb2922d627e33405959dc9bb39fdbaddb24.

Reply:
Still approve :)